File: python/0034.py
import math


class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if not nums:
            return [-1, -1]
        l = 0
        r = len(nums)
        while l + 1 < r:
            m = (l + r) >> 1
            if nums[m] >= target:
                r = m
            elif nums[m] < target:
                l = m
        if nums[l] < target:
            l += 1
        if l >= len(nums) or nums[l] != target:
            return [-1, -1]
        r = len(nums) - 1
        step = 1 << int(math.log2(max(1, r - l)))
        while nums[r] > target and step:
            if r > step and nums[r - step] > target:
                r -= step
            step >>= 1
        if nums[r] > target:
            r -= 1
        return [l, r]


if __name__ == '__main__':
    print(Solution().searchRange([1, 2, 3, 3, 3, 3, 4, 5, 9], 3))

File: rust/src/bin/0034.rs
struct Solution;

fn upper(a: &Vec<i32>, x: i32) -> i32 {
    let mut l = 0;
    let mut r = a.len() - 1;
    while l + 1 < r {
        let mid = (l + r) >> 1;
        if a[mid] > x {
            r = mid - 1;
        } else {
            l = mid;
        }
    }
    if r >= 0 && r < a.len() && a[r] == x {
        r as i32
    } else {
        if l < a.len() && a[l] == x {
            return l as i32;
        }
        -1
    }
}

fn floor(a: &Vec<i32>, x: i32) -> i32 {
    let mut l = 0;
    let mut r = a.len() - 1;
    while l < r {
        let mid = (l + r) >> 1;
        if a[mid] < x {
            l = mid + 1;
        } else {
            r = mid;
        }
    }
    if l >= 0 && l < a.len() && a[l] == x {
        l as i32
    } else {
        if r >= 0 && r < a.len() && a[r] == x {
            return r as i32;
        }
        -1
    }
}

impl Solution {
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        if nums.len() == 0 {
            return vec!(-1, -1);
        }
        return vec!(floor(&nums, target), upper(&nums, target));
    }
}

fn main() {
    let nums = vec![2, 2];
    let target = 0;
    let ans = Solution::search_range(nums, target);
    println!("{:?}", ans);
}